### PR TITLE
Convert result of group by agg to pyarrow if input is pyarrow

### LINKED ISSUE #53030: BUG: Groupby-aggregate on a boolean column returns a different datatype with pyarrow than with numpy
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

#Create a dataframe with a categorical column with two categories and a (numpy) boolean column that is randomly True or False
df = pd.DataFrame.from_dict({'category':['A']*10+['B']*10, 
                             'bool_numpy': np.random.rand(20)>0.5})

#Now make another column that is a copy of the numpy boolean column, but converted to pyarrow
df['bool_arrow'] = df['bool_numpy'].astype('bool[pyarrow]')

print(df.head())
#   category  bool_numpy bool_arrow
# 0        A        True       True
# 1        A        True       True
# 2        A        True       True
# 3        A        True       True
# 4        A       False      False

#Now do a gruopby and aggregate to compute the fraction of True values in each column:
true_fracs = df.groupby('category').agg(lambda x: x.sum()/x.count())

print(true_fracs)

#          bool_numpy bool_arrow
# category                       
# A                0.7       True
# B                0.6       True

#I expect both columns above to have identical floating-point values, not boolean.
```

### Issue Description

Doing a groupby and aggregation on a `bool[pyarrow]` column returns a different datatype than the same operation on a numpy `bool` column. In particular, it seems to always return another `bool[pyarrow]` regardless of the aggregation performed.

### Expected Behavior

I would expect the same datatype and results to be returned regardless of the backend chosen.  Specifically, I would expect the result for category `'A'` to be the same as the result of the following calculation, which *is* the same regardless of backend:
```python
print(df.query("category=='A'")[['bool_numpy','bool_arrow']].sum()/df[['bool_numpy','bool_arrow']].count())
# bool_numpy    0.3
# bool_arrow    0.3
# dtype: float64
```

OR, if this is the intended behavior, I would expect this change to be prominently displayed in the `groupby` documentation.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.8.12.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.0-1032-gcp
Version          : #40~20.04.1-Ubuntu SMP Tue Apr 11 02:49:52 UTC 2023
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.1
numpy            : 1.23.5
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 57.5.0
pip              : 23.0.1
Cython           : 0.29.33
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.10.0
pandas_datareader: None
bs4              : 4.11.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.0
numba            : 0.56.4
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : 2023.1.0
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : 2.3.0
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by undermyumbrella1**
- [ ] closes #53030 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Root cause:
-  `agg_series` always forces output dtype to be the same as input dtype, but depending on the lambda, the output dtype can be different

Fix:
- replace all NA with nan
- convert the `results' to respective pyarrow extension array, using pyarrow library methods
- pyarrow library methods is used instead of `maybe_convert_object`, as `maybe_convert_object` does not check for NA, and forces dtype to float if NA is present (NA is not float in pyarrow),

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 )
 
 import numpy as np
+import pyarrow as pa
```
Comment (mroeschke) [inline pandas/core/groupby/ops.py:21]: pyarrow cannot be imported like this since it's not a required dependency

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 2]
```diff
@@ -1663,3 +1663,319 @@ def func(x):
...
+    expected = DataFrame(
+        {"A": ["cat18", "cat21", "cat39403"], "B": [np.nan, -28.999991, -28.999921]}
+    )
+    expected["B"] = expected["B"].astype("double[pyarrow]")
+    expected.set_index("A", inplace=True)
+    tm.assert_frame_equal(result, expected)
+    assert result["B"].dtype == expected["B"].dtype
+
+
+def test_agg_lambda_pyarrow_to_data_type_conversion():
```
Comment (mroeschke) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1886]: Could you split up these tests per type?

**Review by mroeschke — CHANGES_REQUESTED** [review]
Is this related to a particular issue?

**Comment by undermyumbrella1** [comment]
I have added the issue in the pr, the pr is still work in progress

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -478,6 +478,42 @@ def maybe_cast_pointwise_result(
...
+    try:
+        import pyarrow as pa
+        from pyarrow import (
+            ArrowInvalid,
+            ArrowNotImplementedError,
+        )
+
+        from pandas.core.construction import array as pd_array
+
+        result[isna(result)] = np.nan
```
Comment (rhshadrach) [inline pandas/core/dtypes/cast.py:507]: Why is this necessary?

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -478,6 +478,42 @@ def maybe_cast_pointwise_result(
...
+        pyarrow_result = pa.array(result)
+        pandas_pyarrow_dtype = ArrowDtype(pyarrow_result.type)
+        result = pd_array(result, dtype=pandas_pyarrow_dtype)
+    except (ArrowNotImplementedError, ArrowInvalid):
```
Comment (rhshadrach) [inline pandas/core/dtypes/cast.py:511]: Are these the only exceptions that can be raised by L508-510, or might we run into others?

File: pandas/core/groupby/ops.py [hunk 5]
```diff
@@ -914,18 +916,22 @@ def agg_series(
...
             preserve_dtype = True
 
         result = self._aggregate_series_pure_python(obj, func)
-
         npvalues = lib.maybe_convert_objects(result, try_float=False)
+
         if preserve_dtype:
             out = maybe_cast_pointwise_result(npvalues, obj.dtype, numeric_only=True)
+        elif isinstance(obj._values, ArrowExtensionArray):
+            out = maybe_cast_to_pyarrow_dtype(result, npvalues)
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:934]: If the cast is successful, I think we never use `npvalues`, is that right? In that case, I think we want to avoid the computation that goes into them.

**Review by rhshadrach — COMMENTED** [review]
Thanks for the PR! This appears to me to be a fairly far reaching, and I don't yet feel comfortable given that we have to consider many different cases since the user can provide an arbitrary UDF. It seems to me that the logic "convert to pyarrow dtypes when we can" could result in some surprising behaviors. For example:

```
df = DataFrame({"A": ["c1", "c2", "c3"], "B": [100, 200, 255]})
df["B"] = df["B"].astype("bool[pyarrow]")
gb = df.groupby("A")

result = gb.agg(lambda x: [1, 2, 3])
print(result["B"].dtype)
# list<item: int64>[pyarrow]

result = gb.agg(lambda x: [1, 2, "a"])
print(result["B"].dtype)
# object
```

While I experiment with this some more, a few questions.

File: pandas/core/dtypes/cast.py (hunk 3 above)
Comment (undermyumbrella1) [reply pandas/core/dtypes/cast.py:507]: because pyarrow library is not able to detect NA object when converting from numpy, it can only detect np.nan

**Comment by undermyumbrella1** [comment]
```
df = DataFrame({"A": ["c1", "c2", "c3"], "B": [100, 200, 255]})
df["B"] = df["B"].astype("bool[pyarrow]")
gb = df.groupby("A")

result = gb.agg(lambda x: [1, 2, 3])
print(result["B"].dtype)
# list<item: int64>[pyarrow]

result = gb.agg(lambda x: [1, 2, "a"])
print(result["B"].dtype)
# object
```

for this example, it is expected as that is how pyarrow represents these data structures. E.g homogenous int list and heterogenous object. Alternatively, what would be the expected dtype in this case?

File: pandas/core/groupby/ops.py (hunk 5 above)
Comment (undermyumbrella1) [reply pandas/core/groupby/ops.py:934]: resolved, thanks

File: pandas/core/dtypes/cast.py (hunk 4 above)
Comment (undermyumbrella1) [reply pandas/core/dtypes/cast.py]: resolved, i have included type error and value error raised by `ArrowDtype`, `pd_array` and `pa.array`

**Comment by undermyumbrella1** [comment]
Sorry for the delay in resolving the test cases. Originally the approach is to always caste it back to input dtype, even if the output dtype is different. Currenlty the approach is to let pyarow infer the dtype of the output. However this can result in some unexpected cases as mentioned above. May I check what would be suggested output dtype to fix this issue?

For windows run, it seems to be failing as it is unable to import pyarrow in files unrelated to the code change in the pr, not sure if i have to change some configs.

**Comment by rhshadrach** [comment]
> May I check what would be suggested output dtype to fix this issue?

I think this is the same problem as the discussion here: https://github.com/pandas-dev/pandas/pull/58258#discussion_r1566446371. This appears to me something that will need careful implementation on the EA side.

**Comment by undermyumbrella1** [comment]
Noted, I have made the change according to the comments using `convert_dtypes` and a helper method to account for some unaccounted pyarrow dtypes. Although I am not sure if the change should be made here or in ` lib.maybe_convert_objects`.

File: pandas/core/groupby/ops.py [hunk 6]
```diff
@@ -914,20 +916,26 @@ def agg_series(
         np.ndarray or ExtensionArray
         """
 
-        if not isinstance(obj._values, np.ndarray):
+        result = self._aggregate_series_pure_python(obj, func)
+
+        if isinstance(obj._values, ArrowExtensionArray):
+            return maybe_cast_to_pyarrow_result(result, obj.dtype)
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:922]: Can you use `maybe_cast_pointwise_result` with `same_dtype=False` here instead.

Comment (undermyumbrella1) [reply pandas/core/groupby/ops.py]: resolved

**Comment by undermyumbrella1** [comment]
Thank you for the review, I have updated the pr according to comments.

File: pandas/core/groupby/ops.py [hunk 7]
```diff
@@ -914,20 +916,29 @@ def agg_series(
...
+        if isinstance(obj._values, ArrowExtensionArray):
+            out = maybe_cast_pointwise_result(
+                npvalues, obj.dtype, numeric_only=True, same_dtype=False
+            )
+            import pyarrow as pa
+
+            if isinstance(out.dtype, ArrowDtype) and pa.types.is_struct(
+                out.dtype.pyarrow_dtype
+            ):
+                out = npvalues
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:931]: Is there a test that hits this?

File: pandas/tests/extension/test_arrow.py [hunk 8]
```diff
@@ -225,6 +225,24 @@ def data_for_grouping(dtype):
     return pd.array([B, B, None, None, A, A, B, C], dtype=dtype)
 
 
+def expected_inferred_result_dtype(dtype):
```
Comment (rhshadrach) [inline pandas/tests/extension/test_arrow.py:228]: This depends on the UDF as well as the input dtype - so it could be different for different tests. I think this should be inlined in the one test it's used in instead.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 9]
```diff
@@ -1663,3 +1663,331 @@ def func(x):
     msg = "length must not be 0"
     with pytest.raises(ValueError, match=msg):
         df.groupby("A", observed=False).agg(func)
+
+
+def test_agg_simple_lambda_numpy_to_same_data_type():
+    df = DataFrame(
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1669]: Add GH# for each test - if there isn't a relevant issue, then add this PR's number instead.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 10]
```diff
@@ -1663,3 +1663,331 @@ def func(x):
...
+    )
+    df["B"] = df["B"].astype("bool")
+    gb = df.groupby("A")
+    result = gb.agg(lambda x: x.max())
+
+    expected = DataFrame({"A": [1, 3, 100], "B": [False, False, True]})
+    expected["B"] = expected["B"].astype("bool")
+    expected.set_index("A", inplace=True)
+    tm.assert_frame_equal(result, expected)
+    assert result["B"].dtype == expected["B"].dtype
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1680]: This last assert should not be needed, no? Same for the rest of the tests below.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 11]
```diff
@@ -1663,3 +1663,331 @@ def func(x):
...
+    df = DataFrame(
+        {"A": [1, 3, 100, 3, 100, 100], "B": [False, False, False, False, False, True]}
+    )
+    df["B"] = df["B"].astype("bool")
+    gb = df.groupby("A")
+    result = gb.agg(lambda x: x.max())
+
+    expected = DataFrame({"A": [1, 3, 100], "B": [False, False, True]})
+    expected["B"] = expected["B"].astype("bool")
+    expected.set_index("A", inplace=True)
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1678]: Don't use inplace, it will be removed in the future.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Two general remarks about the tests:

 - Use other pandas methods as little as possible; try to construct what you want directly.
 - It looks like many of the tests added here can be parametrized, can you give that a shot.

For the first, you can do things like

```
df = pd.DataFrame(
    {
        "a": pd.array([1, 2, 3], dtype="..."),
        "b": pd.array([True, False, True], dtype="..."),
    },
    index=pd.Index([1, 2, 3]),
)
```

instead of using `astype`, `set_index`, and the like.

File: pandas/tests/extension/test_arrow.py (hunk 8 above)
Comment (undermyumbrella1) [reply pandas/tests/extension/test_arrow.py]: resolved

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 11 above)
Comment (undermyumbrella1) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: resolved

**Comment by undermyumbrella1** [comment]
Thank you foe the review, I have made changes according to the pr comments.

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (rhshadrach) [reply pandas/core/groupby/ops.py:931]: @jbrockmendel - I was surprised `maybe_cast_pointwise_result` was giving us back a Arrow dtypes we don't have EAs for.  I'm thinking the logic here to prevent this should maybe go in `dtypes.cast._maybe_cast_to_extension_array` in a followup. Any thoughts?

File: pandas/tests/extension/test_arrow.py [hunk 12]
```diff
@@ -1125,6 +1125,38 @@ def test_comp_masked_numpy(self, masked_dtype, comparison_op):
...
+        df = pd.DataFrame({"A": [1, 1, 2, 2, 3, 3, 1, 4], "B": data_for_grouping})
+
+        def expected_inferred_result_dtype(dtype):
+            pa_dtype = dtype.pyarrow_dtype
+            if pa.types.is_date(pa_dtype):
+                return "date32[day][pyarrow]"
+            elif pa.types.is_time(pa_dtype):
+                return "time64[us][pyarrow]"
+            elif pa.types.is_decimal(pa_dtype):
+                return ArrowDtype(pa.decimal128(4, 3))
```
Comment (rhshadrach) [inline pandas/tests/extension/test_arrow.py:1139]: On closer look, I think this is a bug being introduced here. This test is using `.first()`, it should be preserving the dtype in all cases. The changes in this PR now ignore the `preserve_dtype` argument of `agg_series`. When that is true, we should be passing `same_dtype=True` to `maybe_cast_pointwise_result`.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:931]: > giving us back a Arrow dtypes we don't have EAs for

Can you give an example?  this confuses me.  

> should maybe go in dtypes.cast._maybe_cast_to_extension_array

_maybe_cast_to_extension_array is only used in maybe_cast_pointwise_result, so not a huge deal either way.

File: pandas/tests/extension/test_arrow.py (hunk 12 above)
Comment (undermyumbrella1) [reply pandas/tests/extension/test_arrow.py:1139]: resolved, keeping checks for if result is struct, for cases where lambda returns an object and preserve_dtype = false

File: pandas/tests/extension/test_arrow.py [hunk 13]
```diff
@@ -1125,6 +1125,27 @@ def test_comp_masked_numpy(self, masked_dtype, comparison_op):
         expected = pd.Series(exp, dtype=ArrowDtype(pa.bool_()))
         tm.assert_series_equal(result, expected)
 
+    def test_groupby_agg_extension(self, data_for_grouping):
```
Comment (rhshadrach) [inline pandas/tests/extension/test_arrow.py:1128]: I think this test should behave the same as the one in the base class. If that's the case, this can be removed. Can you confirm?

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (rhshadrach) [reply pandas/core/groupby/ops.py:931]:
```python
from pandas.core.dtypes.cast import maybe_cast_pointwise_result

arr = np.array([{"number": 1}])
result = maybe_cast_pointwise_result(
    arr, 
    dtype=pd.ArrowDtype(pa.int64()), 
    numeric_only=True, 
    same_dtype=False,
)
print(result)
# Length: 1, dtype: struct<number: int64>[pyarrow]
```

Comment (rhshadrach) [reply pandas/core/groupby/ops.py:931]: @jbrockmendel - sorry for the noise, I was not aware we could support struct dtypes. I think everything is okay here.

@undermyumbrella1 - why go with NumPy object dtype instead of struct dtypes here?

File: pandas/tests/extension/test_arrow.py (hunk 13 above)
Comment (undermyumbrella1) [reply pandas/tests/extension/test_arrow.py]: resolved

**Comment by rhshadrach** [comment]
I plan to push this across the finish line.

**Comment by undermyumbrella1** [comment]
Ah sorry, I’ll get to it this weekend

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
